src/qibolab/instruments/icarusqfpga.py

```diff
@@ -88,94 +88,99 @@ def play(
-                )
-                sinalpha = np.sin(
-                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
-                )
-                wfm = i_env * sinalpha + q_env * cosalpha
-
-            elif pulse.type == PulseType.READOUT:
-                # For readout pulses, we move the corresponding DAC/ADC pair to the start of the pulse to save memory
-                # This locks the phase of the readout in the demodulation
-                adc = self.ports(pulse.channel).adc
-                start = 0
-
-                end = start + len(i_env)
-                t = np.arange(start, end) / dac_sampling_rate
-                cosalpha = np.cos(
-                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
-                )
-                sinalpha = np.sin(
-                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
-                )
-                wfm = i_env * sinalpha + q_env * cosalpha
-
-                # First we convert the pulse starting time to number of ADC samples
-                # Then, we convert this number to the number of ADC clock cycles (8 samples per clock cycle)
-                # Next, we raise it to the next nearest integer to prevent an overlap between drive and readout pulses
-                # Finally, we ensure that the number is even for the DAC delay conversion
-                delay_start_adc = int(
-                    int(
-                        np.ceil(
-                            self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9 / 8
+        for ch, seq in sequence.items():
+            for pulse in seq:
+                # pylint: disable=no-member
+                # FIXME: ignore complaint about non-existent ports and _ports properties, until we upgrade this driver to qibolab 0.2
+                if pulse.channel not in self._ports:
+                    continue
+
+                dac = self.ports(pulse.channel).dac
+                start = int(pulse.start * 1e-9 * dac_sampling_rate)
+                i_env = pulse.envelope_waveform_i(dac_sr_ghz).data
+                q_env = pulse.envelope_waveform_q(dac_sr_ghz).data
+
+                # Flux pulses
+                # TODO: Add envelope support for flux pulses
+                if "flux" in ch:
+                    wfm = i_env
+                    end = start + len(wfm)
+
+                # Qubit drive microwave signals
+                elif "drive" in ch:
+                    end = start + len(i_env)
+                    t = np.arange(start, end) / dac_sampling_rate
+                    cosalpha = np.cos(
+                        2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                    )
+                    sinalpha = np.sin(
+                        2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                    )
+                    wfm = i_env * sinalpha + q_env * cosalpha
+
+                elif "probe" in ch:
```

Comment:
The instrument should be aware of the channels, so you should be able to work out the pulse types using something like 
```py
isinstance(self.channels[ch].logical_channel, DcChannel)
```
etc.

Obviously this will be postponed to the driver refactoring, but we could add a `FIXME` here as in other similar cases.

Reply:
> Obviously this will be postponed to the driver refactoring, but we could add a `FIXME` here as in other similar cases.

Done in 88df829